Feature/gmoccapy modern icons

PeterMue opened the pull request:
I've created a icon theme for Gmoccapy using the style and partly the icons from googles material design.

The icon theme follows the freedesktop icon theme specifications and therefore integrates well in the gtk ecosystem.

All icons are so called "symbolic" icons, that means they are monochrome icons that automatically get colored according to the theme colors (e.g. in a dark theme the icon color will be white'ish whereas in a light theme the icons will be black'ish). Beside the main color, there are three more possible colors: "error", "warning" and "success" (color value is spedlcified in the theme).
I used these additional colors to mark special accents in the icons (e.g. "error" usually maps to a red'ish color to highlight something disabled or danger)

There is a drop-down to the settings menu to select a custom icon theme.
I've tried to implement this as "minimal invasive" as possible without changing that much code of Gmoccapy and without disrupting the original icons.
(I swap the pixbuf of the Gtk.Image widgets when selecting a theme)

Some screenshots of the new icons with a dark theme (adwaita dark):

![gmoccapy-adwaita-dark-material-auto](https://user-images.githubusercontent.com/3819198/151664395-58d03259-71eb-48c3-b618-f2065474979a.png)
![gmoccapy-adwaita-dark-material-manual](https://user-images.githubusercontent.com/3819198/151664398-8fad3a8f-f4e7-4691-b87c-11001d12a6bd.png)

There is a post with more screenshots at: https://forum.linuxcnc.org/gmoccapy/44925-gmoccapy-icon-theme

With the hope of making a valuable contribution.
Cheers, Peter

c-morley commented:
This branch is based on master-gtk3 which has been merged into master and I think is unused now.
do you want this in master-gtk3 or master branch?

PeterMue commented:
Well, if master-gtk3 is a dead branch it seems pretty useless to merge into, so master is probably better. What do you think?

rene-dev commented:
Master-gtk3 has been merged. Please make a pr to master.

hansu commented:
First, thanks for your work!

## 1. Technical part
1. Why are the SVG-icons placed in a different folder than the PNGs?
2. Whats the need for the `_build_helper.py,` the `build.sh` ans the empty `__init__.py` ? Seems not really be needed to run the icon-theme.
3. Can you move the default Gmoccapy icons from `images` also to `icons/gmoccapy-default/` or sth. like this?
4. And can you handle the Gmoccapy icons also as an icon-theme to not have to restart after switching back to default?
5. Is the 'actions' folder necessary?

6. There is a little bug when switching from material to default and then back to material without restarting in between. Then the the icons loose their dark colour (in case of light theme).

7. Here is something wrong 
   
   ![image](https://user-images.githubusercontent.com/67957916/153751208-19238f15-baf1-4806-92e8-22189c4db146.png)

Except this, good job!

## 2. Design part / usability

I assume you created most of the icons (the machine-related) by yourself. Is that right?

I think the icons have to be optimized a little bit more for a light theme as it is the default theme.
Here some examples (the icons looks like disabled when active):
Maybe its not possible by design, so maybe provide a 'material-dark' and a 'material-light' ?

![image](https://user-images.githubusercontent.com/67957916/153649140-d4301c91-ee9d-491f-9599-15addeb35409.png)  

  
![image](https://user-images.githubusercontent.com/67957916/153750484-8bc00424-5c1c-46ab-8f47-94320f82fa62.png)

Now I give you very detailed feedback about the icon design. It's only what came into my mind first, so please don't feel bothered.

- This icon looks very similar to the exit icon. Little bit irritating for me:
   
   ![image](https://user-images.githubusercontent.com/67957916/153650042-b8c57276-d560-49b5-b31a-48ece9341b37.png)

- Some of your icons are really detailed (like the touch off icons) and some are not (the buttons to the right to switch the mode).
 (Not 100% consistent, even though I like your detailed icons :) and I have not really a proposal)
 Maybe you could remove the 'Set' from this icons, as it is a bit too much on this icon
![image](https://user-images.githubusercontent.com/67957916/153750854-88faf319-4cd5-4412-b581-2b014c8f7bed.png)

- Not really suitable for an emergency stop button: 
![image](https://user-images.githubusercontent.com/67957916/153750982-7e72eb0b-7c73-4689-860e-78943ec1df09.png)

- i don't really link this button. And what should the "L" on all the right buttons express?
![image](https://user-images.githubusercontent.com/67957916/153751010-0f130923-9f3c-46c3-9880-09ad4f1ff85e.png)
I know it is not easy to find a button for this. I liked the old Gmoccapy button with the hand (I think it was in LCNC 2.6)

- Could also be mist instead of flood
![image](https://user-images.githubusercontent.com/67957916/153751171-265c8179-22c7-4362-bf13-7824affd5aa2.png)

hansu commented:
And It would be nice to write a section in the documentation how to create and add a new icon-theme.

hansu commented:
And I think for this image the two states should be swapped.
![image](https://user-images.githubusercontent.com/67957916/153760535-1e2e463a-38ef-4aab-a072-edb325cbad9f.png)
When the machine is off it shows the green icon.

PeterMue commented:
Thank you for the detailed feedback.
I'll appreciate that very much!

I haven't the time to answer the questions right now, but I'll try to do as soon as possible.

PeterMue commented on src/emc/usr_intf/gmoccapy/gmoccapy.py:
```diff
@@ -100,6 +100,9 @@ def excepthook(exc_type, exc_obj, exc_tb):
 _BB_LOAD_FILE = 8
 #_BB_HOME_JOINTS will not be used, we will reorder the notebooks to get the correct page shown
 
+# Default button size for bottom buttons
```
True, but the side buttons are exclusively defined inside the glade file.

hansu replied:
Yes you are right. I thought the bottom ones also.

PeterMue commented on src/emc/usr_intf/gmoccapy/gmoccapy.py:
```diff
@@ -2845,14 +2943,7 @@ def _on_btn_macro_pressed( self, widget = None, data = None ):
-        filepath = os.path.join(IMAGEDIR, file)
-        image = self.macro_dic["keyboard"].get_children()[0]
-        image.set_from_file(filepath)
+        self.macro_dic["keyboard"].set_image(self.widgets[image])
```
Yes, I did, although I really can't understand why it's implemented that way, nor do I understand the use case. But anyway, it behaves the same as before.

hansu replied:
I never installed an on-screen keyboard. Did you?

PeterMue replied:
Yes I did to test the behavior.
What makes me curious is the implementation, if a OSK is installed, the button meant to stop a running macro is redefined to toggle the keyboard. I would have simply added the toggle button additionally if I had implemented it initially. However, not knowing the scenario, I left the functionality exactly as it was.

PeterMue commented:
## 1. Technical

> 1. Why are the SVG-icons placed in a different folder than the PNGs?

Think of the svgs like the source code for the final images.
They contain multiple layers with different icons and/or different icon states (inactive/active for example).
I use them to render the final PNG versions in different sizes and states.
That's where this comes to play:

> 2. Whats the need for the \_build_helper.py, the build.sh ans the empty \_\_init__.py ? Seems not really be needed to run the icon-theme.

These are just build utilities to generate the final icon theme which consists of different sized PNGs (for different screen resolutions and icon sizes), a index file and (in this case a license file).
Theres no runtime dependency, just at build time.
The empty \_\_init__.py is required for python to treat the build helper as module and thus simplifies the import.
Theoretically it's possible to generate the icons when building linuxcnc but thats unnecessary in my opinion and would add more complexity and build dependency (e.g. inkscape) and thus I decided to add the final rendered icon theme to git as well.

> 3. Can you move the default Gmoccapy icons from images also to icons/gmoccapy-default/ or sth. like this?

That's no big deal but imho only makes sense als icon theme (so depends on 4.)

> 4. And can you handle the Gmoccapy icons also as an icon-theme to not have to restart after switching back to default?

Sure that's possible. The reason why I not already did this, is because I wanted to keep the default images as is (they're mostly specified as images in the glade file).
They behave a bit different than symbolic icons and I did not yet find a way to specify a custom GTK icon theme *before* loading the glade file.
That would probably be the best solution as we would be able to just use the icon names regardless if we're in glade or creating images "at runtime". That way, we could delegate the entire icon handling (with recoloring) to GTK.

I'm not sure when I'll find time to implement this and if it's even possible to do so.
Maybe that's worth a future improvement in another PR?

> 5. Is the 'actions' folder necessary?

That's how freedesktop specifies the standard icon theme directory layout, afaik it's not forced but best practice to categorize the icons.
See section "Context" at the [Freedesktop Icon Theme Specification](https://specifications.freedesktop.org/icon-theme-spec/icon-theme-spec-latest.html#Context) (and that's how GTK expects icon themes)

> 6. There is a little bug when switching from material to default and then back to material without restarting in between. Then the the icons loose their dark colour (in case of light theme).

I'll have to look into that, I already have an idea what could be causing it.

> 7. Here is something wrong

I'll have to look into that.

## 2. Design part / usability

> 1. I assume you created most of the icons (the machine-related) by yourself. Is that right?
> I think the icons have to be optimized a little bit more for a light theme as it is the default theme.
> Here some examples (the icons looks like disabled when active):
> Maybe its not possible by design, so maybe provide a 'material-dark' and a 'material-light' ?

You're right, the more basic shaped icons are mostly from the material icon lib, while the more detailed machine specific are made by myself.
I know what you mean and I feel the same way but that's what I finally ended up with. I found it pretty hard to simplify such special use cases which moreover also differ only in tiny details. 

For example the one you've suggested to remove the "set" text. There are two uses for this kind of icon: Set the axis origin to zero and Set the axis origin to a specific value. So the fundamental use case is changing the axis origin but you have to express the two variants "set to zero" and "set to value" in a preferable simple icon - for me, thats pretty hard (i'm a developer not a designer :-P ). I decided to distinguish them with the presence of the "set".

No excuse, just to explain my thoughts.

The other thing is the color handling. That's a feature of GTK and how GTK treats so called symbolic icons. I agree with you, the white color for active state in the light theme is badly recognizable but that's how the GTK theme (The general theme, not icon theme! e.g. adwaita) specified the color.
Although it would be possible to make a biased material-dark/-light theme, it would be exactly that: opinionated for a specific color theme. The idea by using GTK's auto-coloring of symbolic icons is that the icons are more adapting to the themes color schema.

Let me think about this, maybe it's worth to make a biased light variant anyway.

> Now I give you very detailed feedback about the icon design. It's only what came into my mind first, so please don't feel bothered.

It's okay, I'm fine with feedback!
I really appreciate how much effort you out into the review.

> This icon looks very similar to the exit icon. Little bit irritating for me:

You're right, I also came about this. Both icons are from the material design icons and differ between "logout" and "exit to app" and are the closest match to the use case in my opinion.
Maybe a simple "Exit" oder "Quit" text is better? Or do you have a proposal?
(Imho is the presence of the exit button at this prominent location superfluous anyway)

> Some of your icons are really detailed (like the touch off icons) and some are not (the buttons to the right to switch the mode).
> (Not 100% consistent, even though I like your detailed icons :) and I have not really a proposal)
> Maybe you could remove the 'Set' from this icons, as it is a bit too much on this icon

As mentioned above.

> Not really suitable for an emergency stop button:

I agree, I'll find a better icon.

> i don't really link this button. And what should the "L" on all the right buttons express?
> I know it is not easy to find a button for this. I liked the old Gmoccapy button with the hand (I think it was in LCNC 2.6)

It is supposed to represent a control pad, which I think fits well.

The "L" shape is simply a design decision and the "active" indicator as it changes the color when active. Nothing more, just "cause I like it" :-)

> Could also be mist instead of flood

But only viewed in isolation, in conjunction with the fog icon (which is always displayed next to it), I think it is clearly visible. 

> And It would be nice to write a section in the documentation how to create and add a new icon-theme.

I agree with that, I'll write some docs.

> And I think for this image the two states should be swapped.
> When the machine is off it shows the green icon.

I'll have a look at the icon/button in general, as mentioned above, thats not an appropriate emergency stop icon.

hansu commented:
>Think of the svgs like the source code for the final images.
I use them to render the final PNG versions in different sizes and states.

Ah okay that makes sense. I didn't get that.

>Theoretically it's possible to generate the icons when building linuxcnc but thats unnecessary in my opinion and would add more complexity and build dependency (e.g. inkscape) and thus I decided to add the final rendered icon theme to git as well.

I agree. But you can add a target to the makefile that can be run optionally.
Another thought of me is to have the icon sources in a separate repository because they are not really needed to build LinuxCNC. Maybe in your own. Or maybe integrate it as LinuxCNC sub project if some more are added.

>I know what you mean and I feel the same way but that's what I finally ended up with. I found it pretty hard to simplify such special use cases which moreover also differ only in tiny details.

Maybe add another color? On some icons you have some more colors, so they are now not strictly monochrome. But you want let Gtk handle the colors, I understand. Hmm...

>For example the one you've suggested to remove the "set" text. There are two uses for this kind of icon: Set the axis origin to zero and Set the axis origin to a specific value. 

Where is a button "Set the axis origin to zero"? IMHO these buttons are gone in 2.8.

>You're right, I also came about this. Both icons are from the material design icons and differ between "logout" and "exit to app" and are the closest match to the use case in my opinion.
Maybe a simple "Exit" oder "Quit" text is better? Or do you have a proposal?

Maybe something with an arrow from the right to the left? Or just mirror this icon?

c-morley commented:
Where are we at with this please?

hansu commented:
I think (I hope) @PeterMue is doing a very little rework on my suggestions. Peter, when do you think you will have time to finish this?

PeterMue commented:
Yes i will do so, but i can't say when. Maybe by the end of this week.

PeterMue commented:
> 7\. Here is something wrong
>     ![image](https://user-images.githubusercontent.com/67957916/153751208-19238f15-baf1-4806-92e8-22189c4db146.png)

I've fixed the broken icon (there was a typo in the icon name).

The right one is a bit different, there was previously used a gtk stock icon named "gtk-open" [STOCK_OPEN](https://docs.gtk.org/gtk3/const.STOCK_OPEN.html) which is deprecated since GTK 3.10.
The suggested replacement is 

> Use named icon "document-open" or the label "_Open".

So i decided to use document-open, which is what's on the screenshot above.
It's maybe a bit confusing because the same icon is used one level before: 
![image](https://user-images.githubusercontent.com/3819198/157325228-ad9b8599-6a61-48cb-ad5a-a00f4ece44ea.png)

What do you think? Follow the suggestion and keep the icon or change to text "_Open" or different approach?

hansu commented:
Hi Peter, 
thanks for continue working on that. I'll have a look on that later this week.

> The right one is a bit different, there was previously used a gtk stock icon named "gtk-open" [STOCK_OPEN](https://docs.gtk.org/gtk3/const.STOCK_OPEN.html) which is deprecated since GTK 3.10. The suggested replacement is
> 
No, it was `gtk-ok` before. (https://developer-old.gnome.org/gtk2/stable/gtk2-Stock-Items.html#GTK-STOCK-OK:CAPS)

The suggestion here is not very helpful: "Do not use an icon. Use label "_OK"."
I can't find a suitable replacement right now. Maybe you have an idea?

PeterMue commented:
Ah, you're right, it was STOCK_OK and since there's no replacement i searched for STOCK_OPEN (as that's the action of the button).

Well, i think the intention is to open the selected file and therefore i'd rather stick to the document-open icon.
Replacing it with the text "_Open" breaks the visual appearance since everything else is an icon.
I didn't find a similar icon that could replace the old (and has no other semantic bound to).

hansu commented:
In my opinion the intention for this icon is to open a folder like the user has already clicked it to get to this screen. So I think it is confusing to have the same icon here again.

So if we don't want a text label here, I suppose to add the icon to the repository and include it as a image.

hansu commented:
Nice work by the way with the light theme. Looks really good now! :+1:

PeterMue commented:
Now i think i have fixed most of the findings:

- [x] Moved icon theme source files to separate repository (https://github.com/PeterMue/gmoccapy-icon-themes) 
- [x] Add a "classic" icon theme (an thus enabling switch to classic theme without restart)
- [x] Fix wrong icons in classic theme
- [x] Add biased "light" theme
- [x] New "exit" icon that is more different from "back"
- [x] Remove text "set" from "touch-off" icons to have a more simple appearance
- [x] New clear emergency stop button
- [x] Add documentation
- [x] There is a little bug when switching from material to default and then back to material without restarting in between. Then the the icons loose their dark colour (in case of light theme).

The fix for the last finding is imho a bit "blurry". The cause is that the auto coloring needs a style context and i used the style context of the corresponding parent element (usually the button) if available and otherwise default to a blank newly created buttons style context. Now while beeing inside the settings menu (e.g. when switchig the theme) some buttons are disabled (an thus the style context represents "disabled"). Thats the reason why these icons "lost" their right color.

To fix this i changed the code to always use the default style context. There are maybe some circumstances where this is not exactly the right color but at least it's always consistent.

Imho another todo emerged due to the "classic" icon theme: I kept the images folder as these are referenced inside the glade file. I think it would be clean to remove these as well as the references in the glade file.
But i would not do this in this PR.

I would suggest to merge this PR and maybe open a task to cleanup the images.

Cheers,
Peter

hansu commented:
Nice! I like your new file_open icon.
But it seems that something gone wrong when you merged the master in your branch. It says now: Files changed 1,172. 
And a diff from your branch to master confirms this.

One suggestion so far: I would rename the the theme 'material' into 'material-dark' to better distinguish between them.
And a question: What is the purpose of the folder 'scalable'? Is this a kind of a source folder for creating different icon sizes if may needed?

hansu commented:
>Imho another todo emerged due to the "classic" icon theme: I kept the images folder as these are referenced inside the glade file. I think it would be clean to remove these as well as the references in the glade file.
But i would not do this in this PR.

Yes we can remove the images folder later. And the settings menu entry "none". But then we should put the window icon somewhere.

>There is a little bug when switching from material to default and then back to material without restarting in between. Then the the icons loose their dark colour (in case of light theme).

You are meaning this, right?

![Screenshot_2022-03-25_23-26-05](https://user-images.githubusercontent.com/67957916/160210357-7fb2fd0c-69d0-4d63-9dd9-9a98369e6b97.png)

hansu commented:
it is not possible to set a the icon size by the icon theme, or is it?  
Because if I set the correct icon size for the classic theme for these icons, then your material icons will be too big.

![Screenshot_2022-03-25_18-58-29](https://user-images.githubusercontent.com/67957916/160211061-c630c14c-a0fc-4c9a-abb9-c65a57b5288c.png)

hansu commented:
Another thing came into my mind. I think it doesn't make sense to have the icons present in different sizes when they are just the same, only scaled down. Glade can scale them itself. I think we should have available in different sizes when they really differ. What do you think?
For example this icon in different sizes:
16x16 ![document-save](https://user-images.githubusercontent.com/67957916/160233262-aa1b0049-7fd3-4778-ac41-6782c4fb13f6.png)
32x32 ![document-save](https://user-images.githubusercontent.com/67957916/160233271-d66f197a-59f3-4f58-bded-0358beea865d.png)
But no worries, you don't have to deal with that. I can do this :wink:

hansu commented:
And very nice your documentation :+1:  :+1: 
I think we should link somewhere to your repo where your material sources are stored. It would fit well in the NOTICE file I think.

hansu commented:
Hi Peter,
when do you have time to continue on this? Would be nice if we could have it in the 2.9 release.

PeterMue commented:
I think I can do some work tonight, haven't checkt how many todos there are but I assume it's not that much.
What's the deadline for 2.9?

hansu commented:
:+1: 
No hard deadline so far but we are thinking about which PRs should go in master for the 2.9 release and which not.

PeterMue commented:
> Another thing came into my mind. I think it doesn't make sense to have the icons present in different sizes when they are just the same, only scaled down. Glade can scale them itself. I think we should have available in different sizes when they really differ. What do you think?
> For example this icon in different sizes:
> 16x16 ![document-save](https://user-images.githubusercontent.com/67957916/160233262-aa1b0049-7fd3-4778-ac41-6782c4fb13f6.png)
> 32x32 ![document-save](https://user-images.githubusercontent.com/67957916/160233271-d66f197a-59f3-4f58-bded-0358beea865d.png)
> But no worries, you don't have to deal with that. I can do this :wink: 
> 
> 
> 
>  

Sure, that will work also in general.
Only thing that came in my mind is how the scaling in gtk works, I'm not sure if the quality is good in all sizes when scaling down a pixel (png) graphic. At least we have no control over the scaling method (bilinear, cubic, etc).
If we deliver instead pre scaled versions we have full control, and in case of my themes they are direct rendered from a scalable vector graphic.

I don't think that it hurts much to ship them pre scaled, although it sure bloats the package a bit.

hansu commented:
I mean it makes totally sense for your icons which are rendered from an SVG. I was only talking about the original classic icons, sorry. They are kind of a mix of different sizes. Some are in 24x24, some 32x32 and some 48x48. If we derive the other sizes of them, it makes it hard to determine which was the native size (in case we want to make changes). 
So maybe treating these icon sets differently makes sense.

PeterMue commented:
Ah, got it (at least i think so :-D )

Have you seen this: https://github.com/PeterMue/gmoccapy-icon-themes/blob/main/src/classic/build.sh
I copied the original icons there and scale only down if the source is big enough (no upscaling).
So changing the source would be pretty simple as theres only one base icon, wouldn't it?

Maybe we should better integrate the icon theme source files into the linuxcnc repo/build process, so that not the rendered icons are adjusted by mistake but the source. Or make it clear where the icons came from?

c-morley commented:
With no date set it's harder to say - which was the point of my email to list.
To generalize, if code is self contained, then it's easier to allow the risk.
Ie
new hardware driver are almost always included - if they are buggy only that hardware is affected.
Non-critical-failure code such as icons and style code for guis
simply/easily testable component code

Also the more popular the code is the easier bugs are found.
Panelui is not used much, so changes that have bugs will take a long time to show.
AXIS bugs will probably show quicker.

Modern icons is probably fine to add now to master.

Chris
________________________________
From: Hans ***@***.***>
Sent: April 1, 2022 6:12 PM
To: LinuxCNC/linuxcnc ***@***.***>
Cc: c-morley ***@***.***>; Comment ***@***.***>
Subject: Re: [LinuxCNC/linuxcnc] Feature/gmoccapy modern icons (PR #1548)


👍
No hard deadline so far but we are thinking about which PRs should go in master for the 2.9 release and which not.

—
Reply to this email directly, view it on GitHub<https://github.com/LinuxCNC/linuxcnc/pull/1548#issuecomment-1086196444>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ABYGWNO4KEI2BEWOGUSPWSTVC437NANCNFSM5NC6N4GA>.
You are receiving this because you commented.Message ID: ***@***.***>

PeterMue commented:
> A suggestion: I would rename the the theme 'material' into 'material-dark' to better distinguish between light and dark.

No, I wouldn't do that, since it's not an explicit "dark" theme.
(It's a symbolic theme, maybe material-symbolic if we really need to distinguish it - my hope would be that someday ne specialized light/dark themes were needed because the desktop theme does a better coloring job ;-P )

> > There is a little bug when switching from material to default and then back to material without restarting in between. Then the the icons loose their dark colour (in case of light theme).
> 
> You are meaning this, right?
> 
> ![Screenshot_2022-03-25_23-26-05](https://user-images.githubusercontent.com/67957916/160210357-7fb2fd0c-69d0-4d63-9dd9-9a98369e6b97.png)

Yes, exactly.

hansu commented:
>I copied the original icons there and scale only down if the source is big enough (no upscaling).
So changing the source would be pretty simple as theres only one base icon, wouldn't it?

Okay but in the special case of the 36x36 icons: You scaled them down to 32x32. So if they are later maybe used as 48x48, they are scaled up by Gtk and would look worse than before. 
We can have them scaled down that would not hurt. But I see more disadvantages than advantages here:
Pro: 
- Look maybe slightly better when provided in different sizes

Con: 
- More files in the repo
- for some icons I created a version for 32x32 and one for 48x48. But the lost are just scaled down. So it is hard to distinguish which icon is just scaled down from 48x48 and which was manually edited work in the smaller sizes

hansu commented:
>No, I wouldn't do that, since it's not an explicit "dark" theme.

Ok yes, you are right. So just keep the name.

PeterMue commented:
> 

Okay, then i'll accept your changes to classic theme in linxucnc repo and remove the classic from my https://github.com/PeterMue/gmoccapy-icon-themes repo.

hansu commented:
> > 
> 
> Okay, then i'll accept your changes to classic theme in linxucnc repo and remove the classic from my https://github.com/PeterMue/gmoccapy-icon-themes repo.

Good question whether to keep them there or not. Maybe a bit duplicated. I didn't know that you have the classic icons also in [gmoccapy-icon-themes](https://github.com/PeterMue/gmoccapy-icon-themes). 
And I see you have the above mentioned icons there in 36x36 - so I recall my concerns about that - sorry.

PeterMue commented:
Yes it's a useless duplication.

It only makes sense if the classic theme is also some kind of "source" like material with its svg's. Otherwise it's just a 1:1 duplicate.
It's perfectly finde, already removed it from my icon repo (https://github.com/PeterMue/gmoccapy-icon-themes/commit/8a287cbf57292df82638b368fd28e82e91cce2f1)

PeterMue commented:
Are there any to-dos still open that I've overlooked?

hansu commented:
> Are there any to-dos still open that I've overlooked?

Yes that little bug. I thought I have mentioned that, but it seems I haven't, sorry.

>There is a little bug when switching from material to default and then back to material without restarting in between. Then the the icons loose their dark colour (in case of light theme).

The problem is that it doesn't appear only after switching the themes - it appears every time you return from the settings menu until you restart the GUI. So I hope you could have a look at it again.

PeterMue commented:
> > Are there any to-dos still open that I've overlooked?
> 
> Yes that little bug. I thought I have mentioned that, but it seems I haven't, sorry.
> 
> > There is a little bug when switching from material to default and then back to material without restarting in between. Then the the icons loose their dark colour (in case of light theme).
> 
> The problem is that it doesn't appear only after switching the themes - it appears every time you return from the settings menu until you restart the GUI. So I hope you could have a look at it again.

Took me some time to figure out what the problem is, but i think i have at least an idea.
But i think thats a more complicated problem that already existed before but with the new icon behaviour its more obvious.

Haven't yet an idea how to fix it, but I would share my thoughts so far:

The right buttons are as follows:

| Button | Type |
| -------| ----- |
| Estop | ToggleButton |
| On | ToggleButton |
| Mode Manual | RadioButton |
| Mode MDI | RadioButton |
| Mode Auto | RadioButton |
| Settings | ToggleButton |
| UserTabs | ToggleButton |

The RadioButtons are the same group and therefore affect each other.

When toggling "settings" the three mode radio buttons are set to sensitive=False (deactivate), so you can't click them to enter a mode from settings page. (Thats the intended behaviour according to the comments in the code)

But this works only reliable when switching from manual mode to settings.
For both others (MDI, Auto) it's disallowed to be in settings page while these modes are active. But it's allowed to open the settings page because toggling the settings page will automatically switch to manual mode when not already there.

And that's the problem with the radio buttons: when programatically changing the mode to manual, this affects the radio button group states and causes this strange behaviour.
Furthermore, and i think thats a bigger problem at all, it leaves the MDI Mode and Auto Mode buttons active, so you're able to switch back directly to a mode from settings page.

To reproduce on the current 2.8 release:

1. Start Gmoccapy
2. Click Estop
3. Click Machine on
4. Click Settings -> All modes disabled
5. Click Settings -> Manual active (others disable due to not homed)
6. Click Homing menu
7. Click Home all -> All Modes enabled
8. Click Settings -> All Modes disabled
9. Click Settings -> All Modes enabled and in mode manual
10. Click Mode MDI
11. Click Settings -> Only manual disabled others active <-- That's the wrong behaviour imho

See https://youtu.be/Ja1PredSsQw

hansu commented:
>See https://youtu.be/Ja1PredSsQw

Oh yes, I haven't really noticed this, because when you go back from settings everything is fine again :smile: 
I don't know if https://github.com/LinuxCNC/linuxcnc/issues/1543 is may related to this.

>When toggling "settings" the three mode radio buttons are set to sensitive=False (deactivate), so you can't click them to enter a mode from settings page. (Thats the intended behaviour according to the comments in the code)

Yes and IMHO the other buttons don't have to be deactivated when in settings. I could be possible to leave the settings by clicking on another mode button. But yes, it wasn't designed this way and I don't know the specific reason why it was designed like it is now.

>But this works only reliable when switching from manual mode to settings.

Good catch! When entering settings page from manual mode, all is fine. 

>For both others (MDI, Auto) it's disallowed to be in settings page while these modes are active. 

How do you know that ?
What if we drop this restriction? Would that solve this issue? But only as a thought, I am aware that this might have side effects.

>But it's allowed to open the settings page because toggling the settings page will automatically switch to manual mode when not already there.

What would happen if we remember the mode we had before and switch back?

hansu commented:
I think the easiest fix would be to not do this, right?

>When toggling "settings" the three mode radio buttons are set to sensitive=False (deactivate), so you can't click them to enter a mode from settings page.

PeterMue commented:
Right, if we can drop the constraint that the settings page is only allowed while in manual mode, then it's easy.

If not, it's a bit weird because there no actual Viewmodel, instead the state is derived from various ui components in combination with the machine state. I haven't yet a understood how everything works together.

I'll investigate more at the weekend.

Imho before changing such constraints we should understand the reason why it is as it is

hansu commented:
>Imho before changing such constraints we should understand the reason why it is as it is

Yes, I totally agree. I would rather allow clicking the mode radio buttons than dropping such constraints.

Thanks your investigation work! This is really helpful to find the root.

hansu commented:
>Imho before changing such constraints we should understand the reason why it is as it is

It is clearly mentioned here:
```
        # first we set to manual mode, as we do not allow changing settings in other modes
        # otherwise external halui commands could start a program while we are in settings
```

PeterMue commented:
You're totally right, I have read the first comment many times but not yet the second line.

hansu commented:
>When toggling "settings" the three mode radio buttons are set to sensitive=False (deactivate), so you can't click them to enter a mode from settings page. (Thats the intended behaviour according to the comments in the code)

Now I know why this was done. Because when you don't deactivate the mode radio buttons, it will show up as manual mode + settings are active. And the manual mode button becomes active by an event when entering manual mode...
Indeed a bit tricky.

![grafik](https://user-images.githubusercontent.com/67957916/162324542-72db02d9-958f-419f-bde1-c54774d40ece.png)

PeterMue commented:
There's also another special behavior for the user tab button, it's active state is changed by a few different spots. Not sure if it's also involved here.

hansu commented:
I fixed the bug you mentioned (https://github.com/LinuxCNC/linuxcnc/pull/1706) but this didn't solve our problem here.
The problem is the order of execution. 
1. When the setup button is clicked (`on_tbtn_setup_toggled`), it commands a switch to manual mode and `set_sensitive(False)` to the three mode buttons
2. The callback of activating manual mode (`on_hal_status_mode_manual`) sets the manual mode button to active (`set_active()`=pushed). --> This is the point where the button states get confused.
3. The callback of mode change (`on_hal_status_motion_mode_changed()`) sets MDI and Auto button to sensitive(True)

The fix is very simple, I already pushed it to master https://github.com/LinuxCNC/linuxcnc/commit/792bb0dcd10f92c73b754c33055b8db594d30323

So thank you for pointing that out and sorry for bothering you with a bug you didn't introduce.
I think the PR is ready to merge IMHO :tada: . But maybe you would like to test it again with my fix?
I think there is no need to merge the master into your branch as there are no conflicts. But if you feel a bit masochistic, feel free to do a rebase on master to clean up the history a bit :innocent: .

PeterMue commented:
Great! The fix is indeed very simple.
I'll try to rebase this in the next days, not sure when I'll find some time to do.

Thank you!

Cheers

hansu commented:
Thank you. Good teamwork I would say ;-)

hansu commented:
> I'll try to rebase this in the next days, not sure when I'll find some time to do.

Ok, just let me know, when you are finished with this.

PeterMue commented:
Well, I would say I'm done 👍

hansu commented:
I thought that. I don't know exactly what you did, but unfortunately it was not a rebase to LinuxCNC:master.

hansu commented:
Should look something like on https://github.com/hansu/linuxcnc/tree/feature/gmoccapy-modern-icons
But I didn't look at the merge conflicts ans always select "accept incoming". So the result is not correct.

PeterMue commented:
Uhm, sorry about that. I'm not sure whats gone wrong. Indeed the last merge-commit seems to be wrong, isn't it?

hansu commented:
Yeah the last commit looks weird. But the rest seems also not be beginning from master. Have you done 'git rebase master' and fetched master before? 
When I do this a get a merge conflict felt every second commit. So quite some work again and I would understand if you say no to this ;-)

PeterMue commented:
Yeah, i've done this and went through all these merge conflicts.
(Done the rebase using intellij, maybe thats the problem)

hansu commented:
I'm away a couple of days, so will do this later.

hansu commented:
Yeah, looks good now. Thank you!

hansu commented:
So the next steps might be:
- get rid of the images folder at some time
- investigate if we could use the icon theme also in notification.py
